### ISSUE TITLE
Retention policy is applied 1 hour earlier

### ISSUE THREAD

**Issue by EvgeniiSkrebtcov** [corona-warn-app/cwa-server#700]
## Describe the bug

As @Tho-Mat pointed out in https://github.com/corona-warn-app/cwa-server/issues/699#issuecomment-671358818, retention policy applied 1 hour earlier than it should:

> I also noticed that the index for (examples) the hour-file
> 
```
> 2020-07-26-hour-06.zip created 26.07.2020 09:05 (German Time)
> was removed at                 09.08.2020 08:05 (German Time)
> 
> 2020-07-26-hour-08.zip created 26.07.2020 11:05 (German Time)
> was removed at                 09.08.2020 10:05 (German Time)
```
> shouldn't they be removed one hour later to have them 14 Days on the server?

## Expected behavior

Retention policy should apply after 14 days.

**Comment by fredrb** [comment]
The relevant part of the code for retention policy can be found [in this method](https://github.com/corona-warn-app/cwa-server/blob/ee78bf15ece7f282836149b5f39422892f055096/services/distribution/src/main/java/app/coronawarn/server/services/distribution/objectstore/S3RetentionPolicy.java#L60). It deletes all entries with the prefix `version/v1/diagnosis-keys/country/DE/date/<current date>`. This means that all keys for a single day should be deleted together. 

@Tho-Mat could you please share how did you determine the creation and removal dates?

**Comment by Tho-Mat** [comment]
@fredrb 
If you download the hour or date files with "curl -R" you get the creation time/date of the files.
```
curl -R https://svc90.main.px.t-online.de/version/v1/diagnosis-keys/country/DE/date                 --output date.txt
curl -R https://svc90.main.px.t-online.de/version/v1/diagnosis-keys/country/DE/date/2020-08-13/hour --output hour.txt
```

And you are right, the hour-zip-files themselves are delete at the end of the day. 
But the index file "hour.txt" with the active hours "[0,1,2,3,4]" that point to the hour-zip-files  is updated one hour to early.

**Comment by fredrb** [comment]
@Tho-Mat: I'm not sure I fully understand your point. I downloaded all the keys for today to check the creation days, and this was the result.

**Hour files:**
```bash
Aug 13 12:46 .
Aug 13 03:05 0
Aug 13 04:05 1
Aug 13 05:05 2
Aug 13 06:05 3
Aug 13 07:05 4
Aug 13 08:05 5
Aug 13 09:05 6
Aug 13 10:05 7
Aug 13 11:05 8
Aug 13 12:05 9
```
**Index:**
```bash
Aug 13 12:46 .
Aug 13 12:05 hour.txt
```

The index was last updated on the last distribution run at `12:05`. Which I believe it is correct. Am I missing something?
I created [this gist](https://gist.github.com/fredrb/b6e2232ee815d2938f450394da73e76c) with the bash script I made to reproduce this, in case you want to get the same scenario, just run:
```bash
./download_all_day.sh 2020-08-13
```

**Comment by Tho-Mat** [comment]
Ok i think my example for downloading the hour and date index file lead you in the wrong direction.
Today the hour files of 30.07.2020 will be removed. So that is the date I'm talking about.

at 30.07.2020 20:05 German time the hour index file was:
[6,8,9,10,11,13,14,15,17] 
that are the hours that contain keys

**the hour 6 was created at 9:05 German time**
hour files have a time shift of 3 hours (+2 from UTC +1 since hour-0 contains 0:00-01:00)

at 07.08.2020 11:08 German time -due to program change- hour index file contains
[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]

at 12.08.2020 20:05 German time it changes to
[5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 07:05 German time 
[6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 08:05 German time 
[7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]
so the hour 6 was deleted from the index file 
but the hour file was created at 9:05 on 30.07.2020 and **should be remove at 9:05** and not 8:05

at 13.08.2020 09:05 German time 
[8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 10:05 German time 
missing/I don't have
assume:
[9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 11:05 German time 
[10,11,12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 12:05 German time 
[11,12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 13:05 German time
[12,13,14,15,16,17,18,19,20,21,22,23]

at 13.08.2020 14:05 German time
[13,14,15,16,17,18,19,20,21,22,23]

**Comment by fredrb** [comment]
I figured out what is going on. Your analysis is correct, the index file deletes the earliest hour one hour too early. 

The problem that we're facing happens because of the retention policy of the **diagnosis keys**. 

To give an overview of what and why happens:

1. As previously stated in this thread, the retention policy for the **object store** either deletes the whole day or nothing. This means that regardless of the current hour, the files can still be fetched on the CDN. However, in the index file, we can see that entries are removed hourly. And they are removed one hour too soon.
2. The index file is created by the Assembly process, based on the keys mapped to be exported into the hourly files.
3. The mapping of keys is a simple DB read aggregating keys by their submission timestamp. 

During the retention policy step, we delete all the keys from the database that match the query `submission_timestamp<= threshold` where `threshold` is `now - retention_policy`. This means that if we run the retention policy now (14.08 at 14:00), it will delete any entries before or equal (31.07 14:00). Meaning that keys generated for 31.07 14:00 will be deleted, which is not correct, since we are now distributing keys generated between 13:00-14:00 today. 

**To illustrate this better, we can assume a scenario with single day retention policy running distribution day 2 at 8 am.**

### Wrong behavior

| Day 1                | Day 2                 |
| -------------------- | --------------------- |
|                      | 00:00                 |
|                      | 01:00                 |
|                      | 02:00                 |
|                      | 03:00                 |
|                      | 04:00                 |
|                      | 05:00                 |
|                      | 06:00                 |
|  | **07:00** --> Created |
| **08:00** --> Deleted                |                       |
| 09:00                |                       |
| 10:00                |                       |
| 11:00                |                       |
| 12:00                |                       |
| 13:00                |                       |
| 14:00                |                       |
| 15:00                |                       |
| 16:00                |                       |
| 17:00                |                       |
| 18:00                |                       |
| 19:00                |                       |
| 20:00                |                       |
| 21:00                |                       |
| 22:00                |                       |
| 23:00                |                       |

This is wrong because we are only holding 23 hours worth of data. Hour 7 is created at 8:00 but removed at 7:00 the next day. 

### Correct behavior

| Day 1                | Day 2                 |
| -------------------- | --------------------- |
|                      | 00:00                 |
|                      | 01:00                 |
|                      | 02:00                 |
|                      | 03:00                 |
|                      | 04:00                 |
|                      | 05:00                 |
|                      | 06:00                 |
| **07:00** --> Delete | **07:00** --> Created |
| 08:00                |                       |
| 09:00                |                       |
| 10:00                |                       |
| 11:00                |                       |
| 12:00                |                       |
| 13:00                |                       |
| 14:00                |                       |
| 15:00                |                       |
| 16:00                |                       |
| 17:00                |                       |
| 18:00                |                       |
| 19:00                |                       |
| 20:00                |                       |
| 21:00                |                       |
| 22:00                |                       |
| 23:00                |                       |

We should delete everything **before 8am (not inclusive)**, to assure that we will always have 24h of keys in the system. 

There's already a fix for that, will open the PR in a couple of minutes. @Tho-Mat would also appreciate your comments there. Thanks again for your help investigating and explaining this issue.

**Comment by Tho-Mat** [comment]
How long will it take to fix this on the live cwa server?

**Comment by Tho-Mat** [comment]
@fredrb 
10 days after the fix the bug is still active.

**Comment by fredrb** [comment]
> How long will it take to fix this on the live cwa server?

Since this wasn't patched as a hotfix, it wasn't in the 1.3 release. As of now it's planned for 1.4 release which will _likely_ be out in a couple of weeks. 

> 10 days after the fix the bug is still active.

We tested in our internal system and it seems to be fixed. But we'll keep the issue open in GitHub until we get a confirmation that it's working on the national servers.

**Comment by fredrb** [comment]
Update from our side: release 1.4 is currently planned for 30.09.2020.

**Comment by Tho-Mat** [comment]
@fredrb 
Thanks for  updating the news.

**Comment by Tho-Mat** [comment]
@EvgeniiSkrebtcov 

**Unexpected behavior**

At 01.10.2020 00:05 German time the hour file for 17.09.2020 contains 
[3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]      
that indicates the fist time a user uploads keys on 17.09.2020 was in the UTC interval 03:00-04:00.     
So the next retention run is expected at 05:00 German time on 01.10.2020.          

But the next run was earlier.
At 01.10.2020 03:00 German time and the file then contains     
[5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]     
The run-time 03:00 German time indicates that some keys were also uploaded in the UTC interval 01:00-02:00.     
The now missing 3,4 indicate there was no key-upload during UTC interval 03:00-04:00 and 04:00-05:00

This is at odds with the first observation.

Is this caused by release 1.4.
But I do not see any changes in the retention policy yet, so i assume old version is still active(or the fix does not work).

Is there a way to see what version of the server is online?

**Comment by fredrb** [comment]
@Tho-Mat: 1.4 is not yet released.

**Comment by Tho-Mat** [comment]
Since 
https://github.com/corona-warn-app/cwa-server/issues/700#issuecomment-702069110
has nothing to do with server version change i will open a new issue

**Comment by EvgeniiSkrebtcov** [comment]
Fix for this issue was shipped with the release of CWA server v1.5
